### PR TITLE
UI: Enable TurboSnap for UI Storybook

### PR DESCRIPTION
**Pull request by ethriel3695**
Issue:

Storybook is using up quite a bit of its story allotment within Chromatic.
We use Vite to build Storybook, which does not work out of the box with the Chromatic feature `TurboSnap`

Add a package to allow the Storybook monorepo to use `TurboSnap`

## What I did

- Added the package `vite-plugin-turbosnap`
- Updated the `main.ts` Storybook config file to include the package

## How to test

- [ ] Is this testable with Jest or Chromatic screenshots?
- [ ] Does this need a new example in the kitchen sink apps?
- [ ] Does this need an update to the documentation?

If your answer is yes to any of these, please make sure to include it in your PR.

### CODE REVIEW

File: code/tsconfig.json [hunk 1]
```diff
@@ -24,5 +24,11 @@
     "compilerOptions": {
       "types": ["node"]
     }
+  },
+  "angularCompilerOptions": {
+    "enableI18nLegacyMessageIdFormat": false,
+    "strictInjectionParameters": true,
+    "strictTemplates": true
+    // ...
```
Comment (JReinhold) [inline code/tsconfig.json:32]: is this needed?

**Review by JReinhold — COMMENTED** [review]
Thanks Reuben!

We need to change the CircleCI config though, as it doesn't contain any Git history. From the CLI output:

```
Retrieving git information
⚠ TurboSnap disabled due to missing git history
Could not retrieve changed files since baseline commit(s).
This typically happens after rebasing, force pushing, or when running against an ephemeral merge commit.
ℹ Read more at https://www.chromatic.com/docs/turbosnap#how-it-works
```

Maybe you can reach out to Chromatic Support and get this resolved? 🤣

**Review by JReinhold — CHANGES_REQUESTED** [review]
.

File: code/tsconfig.json (hunk 1 above)
Comment (ethriel3695) [reply code/tsconfig.json]: Thank you for catching this, I removed the accidentally added settings.

**Comment by ethriel3695** [comment]
> Thanks Reuben!
> 
> We need to change the CircleCI config though, as it doesn't contain any Git history. From the CLI output:
> 
> ```
> Retrieving git information
> ⚠ TurboSnap disabled due to missing git history
> Could not retrieve changed files since baseline commit(s).
> This typically happens after rebasing, force pushing, or when running against an ephemeral merge commit.
> ℹ Read more at https://www.chromatic.com/docs/turbosnap#how-it-works
> ```
> 
> Maybe you can reach out to Chromatic Support and get this resolved? 🤣

I've updated the depth to 5 for the `git-shallow-clone` orb that we use for `git checkout`
https://circleci.com/developer/orbs/orb/guitarrapc/git-shallow-clone

If that does not work, we may want to consider altering this job to utilize the CIrcleCI helper `checkout` instead.

**Comment by ethriel3695** [comment]
@JReinhold @tmeasday 
It's interesting that the CI process indicates that TurboSnap was disabled because we could not find the `git` history
https://app.circleci.com/pipelines/github/storybookjs/storybook/31499/workflows/d5673531-0632-4f9a-b824-32c448749599/jobs/437615

Yet, the build in Chromatic clearly indicates that TurboSnap was successful and skipped snapshots as intended.
Screen Shot 2022-11-07 at 6 11 52 PM

File: scripts/tasks/chromatic.ts [hunk 2]
```diff
@@ -15,6 +15,9 @@ export const chromatic: Task = {
     await exec(
       `npx chromatic \
           --exit-zero-on-changes \
+          --exit-once-uploaded \
+          --only-changed \
+          --debug \
```
Comment (tmeasday) [inline scripts/tasks/chromatic.ts:20]: This is only necessary if we want to enable TS for the sandbox SBs, which I don't believe is the aim of this PR. The `ui:` SBs already have `--only-changed` turned on it seems: https://github.com/storybookjs/storybook/blob/737d3399272343a76a1cd17832af70671ea4c26c/code/package.json#L84

File: .circleci/config.yml [hunk 3]
```diff
@@ -316,8 +316,8 @@ jobs:
         default: 9
       cadence:
         type: enum
-        enum: [ "ci", "daily", "weekly" ]
-        default: "ci"
+        enum: ['ci', 'daily', 'weekly']
```
Comment (tmeasday) [inline .circleci/config.yml:319]: Can you revert these formatting changes?

**Review by tmeasday — CHANGES_REQUESTED** [review]
This is good, with a few tweaks.

> It's interesting that the CI process indicates that TurboSnap was disabled because we could not find the git history

I think the message here is a little confusing, the issue is that snapshotting is entirely disabled for the `storybook-blocks` SB right now: https://www.chromatic.com/manage?appId=63591cc9eaf4e8caec2bb5e9

> Yet, the build in Chromatic clearly indicates that TurboSnap was successful and skipped snapshots as intended.

You're mixing the two SBs that are tested in that command up, that's the `storybook-ui` SB which is working fine it seems.

> The change runs correctly on storybook:ui however, not for the chromatic-sandboxes job

> I've modified the chromatic.ts task and added --only-changed to see if this might resolve the issue

This is a bit trickier, as we would have to inject the vite config into each sandbox. It can be done, but we are looking to heavily reduce the number of sandboxes we run soon anyway. Otherwise we can take a task to do this on the SB team next cycle.

**Comment by JReinhold** [comment]
Yeah sorry for the confusion @ethriel3695 
Just to make it clear, we have 3 types of sandboxes running:

1. The UI Storybook - `storybook:ui:chromatic`, which contains all the manager and block components. This is the one we want TurboSnap for
2. The Blocks Storybook - `storybook:blocks:chromatic`. It is just a subset of the UI Storybook that is for public consumption, and we've disabled UI Tests and UI Reviews for this, it's only for publishing purposes - so TurboSnap is irrelevant. Maybe this is what the CLI is trying to tell us, but with a weird error message?
3. Sandboxes. @tmeasday I don't think we can ever use TurboSnap for these, because they are not checked into git, so there will never be a git history for them? Even if we checked in their webpack stats file I assume that wouldn't be enough, because TS still looks at changes to `package.json`, `yarn.lock`, etc.

1. and 2. both run in the same CI job outside of the sandboxing tasks.

**Comment by tmeasday** [comment]
@JReinhold you might want to change the `storybook:blocks:chromatic` script to not pass `--only-changed` then. Or we can just live with the warning 🤷 

> Sandboxes. @tmeasday I don't think we can ever use TurboSnap for these, because they are not checked into git, so there will never be a git history for them? Even if we checked in their webpack stats file I assume that wouldn't be enough, because TS still looks at changes to package.json, yarn.lock, etc.

Well, hmm. They are in the same git repo. So they will have git changes. But because of the way they are built I don't think TS will work, as all the file changes will be in `node_modules`. So certainly, I don't think we should use TS for them.

@ethriel3695 as I mentioned we are reducing the number of sandboxes we run by 2-3x currently, FYI.

**Review by JReinhold — APPROVED** [review]
Changes look good! 💪 
I don't know why CI is failing though, some weird Chromatic CLI error.
Could be because you updated the Chromatic version?

Could you try merging `next` in and see if that fixed it?

File: .circleci/config.yml (hunk 3 above)
Comment (ethriel3695) [reply .circleci/config.yml]: Reverted the formatting changes

**Comment by IanVS** [comment]
Let me know if `vite-plugin-turbosnap` is ever misbehaving (I wrote it).

**Review by tmeasday — APPROVED** [review]
🙌

**Pull request merged**